File: packages/ts/src/frameworks.ts
/** Framework & testing-tool fingerprinting: match dependency pools against the matrix rules. */
import type { FrameworkResult, TestingResult } from "./types.js";
import type { SignalMatrix } from "./matrix.js";
import type { DependencyPools } from "./parsers.js";

function poolFor(ecosystem: string, pools: DependencyPools): Set<string> {
  switch (ecosystem) {
    case "node":
    case "typescript":
      return pools.js;
    case "python":
      return pools.py;
    case "java-maven":
    case "java-gradle":
    case "kotlin":
      return pools.java;
    case "go":
      return pools.go;
    case "rust":
      return pools.rust;
    default:
      return new Set();
  }
}

export function matchFrameworks(matrix: SignalMatrix, pools: DependencyPools): FrameworkResult[] {
  const out: FrameworkResult[] = [];
  for (const def of matrix.frameworks) {
    const pool = poolFor(def.ecosystem, pools);
    const evidence = def.deps.filter((d) => pool.has(d)).sort();
    if (evidence.length > 0) {
      out.push({ ecosystem: def.ecosystem, name: def.name, evidence });
    }
  }
  return out;
}

export function matchTesting(matrix: SignalMatrix, pools: DependencyPools): TestingResult[] {
  const out: TestingResult[] = [];
  for (const def of matrix.testing) {
    const pool = poolFor(def.ecosystem, pools);
    if (def.deps.some((d) => pool.has(d))) {
      out.push({ ecosystem: def.ecosystem, framework: def.framework });
    }
  }
  return out;
}

File: packages/py/src/repo_fingerprint/frameworks.py
"""Framework & testing-tool fingerprinting from dependency pools."""
from __future__ import annotations

from typing import Any

from .parsers import DependencyPools
from .types import FrameworkResult, TestingResult


def _pool_for(ecosystem: str, pools: DependencyPools) -> set[str]:
    if ecosystem in ("node", "typescript"):
        return pools.js
    if ecosystem == "python":
        return pools.py
    if ecosystem in ("java-maven", "java-gradle", "kotlin"):
        return pools.java
    if ecosystem == "go":
        return pools.go
    if ecosystem == "rust":
        return pools.rust
    return set()


def match_frameworks(matrix: dict[str, Any], pools: DependencyPools) -> list[FrameworkResult]:
    out: list[FrameworkResult] = []
    for defn in matrix["frameworks"]:
        pool = _pool_for(defn["ecosystem"], pools)
        evidence = sorted(d for d in defn["deps"] if d in pool)
        if evidence:
            out.append(FrameworkResult(ecosystem=defn["ecosystem"], name=defn["name"], evidence=evidence))
    return out


def match_testing(matrix: dict[str, Any], pools: DependencyPools) -> list[TestingResult]:
    out: list[TestingResult] = []
    for defn in matrix["testing"]:
        pool = _pool_for(defn["ecosystem"], pools)
        if any(d in pool for d in defn["deps"]):
            out.append(TestingResult(ecosystem=defn["ecosystem"], framework=defn["framework"]))
    return out
